File: radical-red-web/scripts/fetch-rr-moves.js
/**
 * Fetches Radical Red move data from the source repository
 * and creates a moves-data.json file with correct IDs
 */

import fs from 'fs';
import https from 'https';

const MOVES_H_URL = 'https://raw.githubusercontent.com/Klendy/radical_red/master/include/constants/moves.h';
const BATTLE_MOVES_URL = 'https://raw.githubusercontent.com/Klendy/radical_red/master/src/Tables/battle_moves.c';

async function fetchFile(url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
    }).on('error', reject);
  });
}

function parseMovesH(content) {
  const moves = {};
  const lines = content.split('\n');

  for (const line of lines) {
    // Match lines like: #define MOVE_POUND 0x1
    const match = line.match(/^#define\s+MOVE_(\w+)\s+(0x[0-9A-Fa-f]+|\d+)/);
    if (match) {
      const name = match[1];
      const id = match[2].startsWith('0x')
        ? parseInt(match[2], 16)
        : parseInt(match[2], 10);

      // Skip MAX and GMAX moves as they're calculated values
      if (!name.includes('MOVES_COUNT') && !name.includes('_MAX_') && !name.includes('_GMAX_')) {
        // Convert MOVE_NAME to proper case: MOVE_POUND -> Pound
        const properName = name
          .split('_')
          .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
          .join(' ');

        moves[id] = {
          id,
          name: properName,
          constantName: `MOVE_${name}`
        };
      }
    }
  }

  return moves;
}

function parseBattleMovesC(content, moves) {
  const lines = content.split('\n');
  let currentMove = null;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();

    // Match move entry start: [MOVE_POUND] =
    const moveMatch = line.match(/\[MOVE_(\w+)\]\s*=/);
    if (moveMatch) {
      const moveName = moveMatch[1];
      const constantName = `MOVE_${moveName}`;

      // Find the move ID from our parsed moves
      const moveEntry = Object.values(moves).find(m => m.constantName === constantName);
      if (moveEntry) {
        currentMove = moveEntry.id;
      }
      continue;
    }

    if (currentMove !== null && moves[currentMove]) {
      // Extract move properties
      if (line.includes('.power')) {
        const match = line.match(/\.power\s*=\s*(\d+)/);
        if (match) moves[currentMove].power = parseInt(match[1]);
      }

      if (line.includes('.type')) {
        const match = line.match(/\.type\s*=\s*TYPE_(\w+)/);
        if (match) moves[currentMove].type = match[1].toLowerCase();
      }

      if (line.includes('.accuracy')) {
        const match = line.match(/\.accuracy\s*=\s*(\d+)/);
        if (match) moves[currentMove].accuracy = parseInt(match[1]);
      }

      if (line.includes('.pp')) {
        const match = line.match(/\.pp\s*=\s*(\d+)/);
        if (match) moves[currentMove].pp = parseInt(match[1]);
      }

      if (line.includes('.split')) {
        const match = line.match(/\.split\s*=\s*SPLIT_(\w+)/);
        if (match) moves[currentMove].class = match[1].toLowerCase();
      }

      // End of move entry
      if (line.includes('},')) {
        currentMove = null;
      }
    }
  }

  return moves;
}

async function main() {
  try {
    console.log('Fetching Radical Red move constants...');
    const movesH = await fetchFile(MOVES_H_URL);

    console.log('Parsing move IDs and names...');
    let moves = parseMovesH(movesH);

    console.log(`Found ${Object.keys(moves).length} moves`);

    console.log('Fetching Radical Red move data...');
    const battleMovesC = await fetchFile(BATTLE_MOVES_URL);

    console.log('Parsing move properties...');
    moves = parseBattleMovesC(battleMovesC, moves);

    // Convert to the format expected by the app: object with string keys
    const movesData = {};
    Object.values(moves).forEach(move => {
      movesData[move.id] = {
        id: move.id,
        name: move.name,
        type: move.type || 'normal',
        power: move.power || 0,
        accuracy: move.accuracy || 0,
        pp: move.pp || 0,
        class: move.class || 'status'
      };
    });

    // Write to file
    const outputPath = './src/lib/moves-data.json';
    fs.writeFileSync(outputPath, JSON.stringify(movesData, null, 2));

    console.log(`\n✓ Successfully created ${outputPath}`);
    console.log(`✓ Total moves: ${Object.keys(movesData).length}`);

    // Verify the problematic moves
    console.log('\nVerifying key moves:');
    const testMoves = {
      'Black Hole Eclipse (P)': 0x304,
      'Circle Throw': 0x1BD,
      'Reflect Type': 0x28E,
      'Spacial Rend': 0x1ED,
      'Hyperspace Hole': 0x1FD
    };

    for (const [name, id] of Object.entries(testMoves)) {
      const move = movesData[id];
      console.log(`  ${name} (ID ${id}): ${move ? move.name : 'NOT FOUND'}`);
    }

  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

main();

File: radical-red-web/scripts/fetch-rr-moves.py
#!/usr/bin/env python3
"""
Fetches Radical Red move data from the source repository
and creates a moves-data.json file with correct IDs
"""

import json
import re
import urllib.request

MOVES_H_URL = 'https://raw.githubusercontent.com/Klendy/radical_red/master/include/constants/moves.h'

def fetch_file(url):
    """Fetch file content from URL"""
    with urllib.request.urlopen(url) as response:
        return response.read().decode('utf-8')

def parse_moves_h(content):
    """Parse moves.h file to extract move IDs and names"""
    moves = {}
    lines = content.split('\n')

    for line in lines:
        # Match lines like: #define MOVE_POUND 0x1
        match = re.match(r'^#define\s+MOVE_(\w+)\s+(0x[0-9A-Fa-f]+|\d+)', line)
        if match:
            name = match.group(1)
            id_str = match.group(2)

            # Convert to integer
            move_id = int(id_str, 16) if id_str.startswith('0x') else int(id_str)

            # Skip calculated values and special constants
            if 'MOVES_COUNT' in name or '_MAX_' in name or '_GMAX_' in name or name == 'NONE':
                continue

            # Convert MOVE_NAME to proper case
            # Handle special cases first
            if name.endswith('_P'):
                proper_name = format_move_name(name[:-2]) + ' (Physical)'
            elif name.endswith('_S'):
                proper_name = format_move_name(name[:-2]) + ' (Special)'
            else:
                proper_name = format_move_name(name)

            moves[str(move_id)] = {
                'id': move_id,
                'name': proper_name,
                'type': 'normal',  # Default values
                'power': 0,
                'accuracy': 0,
                'pp': 0,
                'class': 'status'
            }

    return moves

def format_move_name(name):
    """Format move name from constant to proper case"""
    # Handle special cases
    special_cases = {
        'VISE': 'Vise',
        'VICEGRIP': 'Vise Grip',
        'HIDDENPOWER': 'Hidden Power',
        'SOLARBEAM': 'Solar Beam',
        'DOUBLEEDGE': 'Double-Edge',
        'SOFTBOILED': 'Soft-Boiled',
        'SELFDESTRUCT': 'Self-Destruct',
        'SMELLINGSALT': 'Smelling Salts',
        'FEINTATTACK': 'Feint Attack',
        'ANCIENTPOWER': 'Ancient Power',
        'EXTREMESPEED': 'Extreme Speed',
        'SHADOWBALL': 'Shadow Ball',
        'FUTURESIGHT': 'Future Sight',
        'ROCKSMASH': 'Rock Smash',
        'FAKEOUT': 'Fake Out',
        'BLAZEKICK': 'Blaze Kick',
        'MUDSPORT': 'Mud Sport',
        'WATERSPORT': 'Water Sport',
        'GRASSWHISTLE': 'Grass Whistle',
        'POISONFANG': 'Poison Fang',
        'MUDSHOT': 'Mud Shot',
        'ROCKBLAST': 'Rock Blast',
        'SHOCKWAVE': 'Shock Wave',
        'MAGICCOAT': 'Magic Coat',
        'ASTONISH': 'Astonish',
        'WEATHERBALL': 'Weather Ball',
        'GRASSYTERRAIN': 'Grassy Terrain',
        'MISTYTERRAIN': 'Misty Terrain',
        'ELECTRICTERRAIN': 'Electric Terrain',
        'PSYCHICTERRAIN': 'Psychic Terrain',
        'HYPERSPACEHOLE': 'Hyperspace Hole',
        'SPACIALREND': 'Spacial Rend',
        'BLACKHOLEECLIPSE': 'Black Hole Eclipse',
        'CIRCLETHROW': 'Circle Throw',
        'REFLECTTYPE': 'Reflect Type',
    }

    if name in special_cases:
        return special_cases[name]

    # Simple approach: capitalize and replace underscores with spaces
    return name.replace('_', ' ').title()

def main():
    try:
        print('Fetching Radical Red move constants...')
        moves_h = fetch_file(MOVES_H_URL)

        print('Parsing move IDs and names...')
        moves = parse_moves_h(moves_h)

        print(f'Found {len(moves)} moves')

        # Write to file
        output_path = './src/lib/moves-data.json'
        with open(output_path, 'w') as f:
            json.dump(moves, f, indent=2)

        print(f'\n✓ Successfully created {output_path}')
        print(f'✓ Total moves: {len(moves)}')

        # Verify the problematic moves
        print('\nVerifying key moves:')
        test_moves = {
            'Black Hole Eclipse (P)': 0x304,
            'Circle Throw': 0x1BD,
            'Reflect Type': 0x28E,
            'Spacial Rend': 0x1ED,
            'Hyperspace Hole': 0x1FD
        }

        for name, move_id in test_moves.items():
            move = moves.get(str(move_id))
            if move:
                print(f'  {name} (ID {move_id}): {move["name"]}')
            else:
                print(f'  {name} (ID {move_id}): NOT FOUND')

    except Exception as e:
        print(f'Error: {e}')
        import traceback
        traceback.print_exc()
        exit(1)

if __name__ == '__main__':
    main()
